### PR TITLE
Use integer tensors for indices for rows, cols, cross_entropy

### PR DESCRIPTION
**Pull request by emjotde**
This PR introduces a couple of clean-up steps and is a first attempt at using the new type system in a more systematic way. This will be further extended to handle fp16 on the GPU and SSE on the CPU. Adding this uncovered a couple of non-lethal bugs.

Changes:
* Use uint32_t everywhere for vocabulary indices and indices in general. We are unlikely to exceed 4 billion indices with our neural models. This reduces copying a bit, also can be changed with changing only the typedef. 
* Introduce ExpressionGraph::indices(std::vector<IndexType>) as an alias to ExpressionGraph::constant(...) for integer tensors to be used as indices.
* Fix bug with CPU `softmax` by removing mask from the code, just use log-mask everywhere similar to `transformer.h`
* Add some run-time type-checking there will be more of this in the future when we start rewriting kernels. 
* Unrelated: also added a new parameter to Adam, which allows to use an Adam version called AdamW. To be evaluated. 
* Remove batch-flexible learning-rate, didn't seem to do anything.

### CODE REVIEW

**Comment by frankseide** [comment]
Few comments, based on the description (have not had time to read the code):

 * Please make sure that the calls remain backwards compatible for a while. Master is a bit too much of a moving target these days. I am worried I will not be able to merge it back with my research code.
 * Why int32_t and not size_t? That should be the natural type.

**Comment by emjotde** [comment]
* I am happy to do the merge for you. I am doing a lot of clean-up here, so some problems are expected to occur. Just let me know if you don't want to deal with this. We can also do a few first merges on Monday to make sure you are not too far behind. There are no functionality changes. Backwards-compatibility for your code should be mostly ensured. If not, it's most likely a mistake and I will fix it. 
* As for uint32, it's actually a bit faster since it's half the size. Changing it back is easy enough now, one typedef, and adjust the runtime-type (which should have a special index type too).

File: src/tensors/cpu/tensor_operators.cpp [hunk 1]
```diff
@@ -707,7 +719,7 @@ void CrossEntropyPickBackward(Tensor out_,
 
     // cross-entropy
     for(int i = 0; i < cols; ++i) {
-      float sub = (float)(i == (int)pick[j]);
+      float sub = (float)(i == (int)pick_->data<IndexType>()[j]);
```
Comment (frankseide) [inline src/tensors/cpu/tensor_operators.cpp]: Here it really looks like `pick_` is a class member.

File: src/tensors/gpu/tensor_operators.cu [hunk 2]
```diff
@@ -1246,7 +1175,9 @@ __global__ void gCrossEntropyPick(float* out,
   }
 }
 
-void CrossEntropyPick(Tensor out, Tensor in, Tensor pick) {
+void CrossEntropyPick(Tensor out, Tensor in, Tensor indices) {
```
Comment (frankseide) [inline src/tensors/gpu/tensor_operators.cu]: From the name I cannot tell what this does. Add a very brief comment?

File: src/training/graph_group_singleton.cpp [hunk 3]
```diff
@@ -18,12 +18,7 @@ void SingletonGraph::execute(Ptr<data::Batch> batch) {
 
   // Get batch stats
   size_t batch_words = batch->wordsTrg();
-
-  if(scaleLearningRate_) {
-    opt_->update(graph_, batch_words / avgBatchWords_);
-  } else {
-    opt_->update(graph_);
-  }
+  opt_->update(graph_);
```
Comment (frankseide) [inline src/training/graph_group_singleton.cpp]: How come I don't see this change for graph_group_sync.cpp? I like that, since I am modifying that file :)

File: src/examples/mnist/dataset.h [hunk 4]
```diff
@@ -17,6 +17,116 @@
 namespace marian {
 namespace data {
 
+typedef std::vector<float> Data;
+typedef std::vector<IndexType> Labels;
+typedef std::vector<Data> Example;
+typedef std::vector<Example> Examples;
+
+typedef Examples::const_iterator ExampleIterator;
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/examples/mnist/dataset.h]: Moved. This was only used for the MNIST example and cluttering the general code.

File: src/examples/mnist/dataset.h [hunk 5]
```diff
@@ -17,6 +17,116 @@
...
+
+class Input {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/examples/mnist/dataset.h]: This is for the MNIST example which I am actually thinking of removing. It's way overcomplicated.

File: src/examples/mnist/model.h [hunk 6]
```diff
@@ -24,9 +24,11 @@ class MNISTCrossEntropyCost : public CostBase {
...
     auto top = model->build(graph, batch, clearGraph);
 
-    auto vLabels = std::static_pointer_cast<data::DataBatch>(batch)->labels();
-    auto labels
-        = graph->constant({(int)batch->size(), 1}, inits::from_vector(vLabels));
+    auto vfLabels = std::static_pointer_cast<data::DataBatch>(batch)->labels();
+
+    // convert float to IndexType
+    std::vector<IndexType> vLabels(vfLabels.begin(), vfLabels.end());
+    auto labels = graph->indices(vLabels);
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/examples/mnist/model.h]: Yes, that is a good idea, added the TODO in ExpressionGraph

File: src/graph/expression_graph.h [hunk 7]
```diff
@@ -347,16 +347,24 @@ class ExpressionGraph : public std::enable_shared_from_this<ExpressionGraph> {
...
 
-  Expr constant(const Shape& shape, const NodeInitializer& init) {
-    return Expression<ConstantNode>(shared_from_this(), shape, init);
+  Expr constant(const Shape& shape, const NodeInitializer& init, Type value_type = Type::float32) {
+    return Expression<ConstantNode>(shared_from_this(), shape, init, value_type);
+  }
+
+  // shortcut to turn vector of indices to integer tensor, to be used with operators
+  // like rows or select
+  Expr indices(const std::vector<IndexType>& indices) {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_graph.h]: done

File: src/graph/expression_operators.cpp [hunk 8]
```diff
@@ -55,8 +55,15 @@ Expr operator-(Expr a) {
...
 
+Expr softmax(Expr a) {
+  return Expression<SoftmaxNodeOp>(a);
+}
+
+// @TODO: maybe get rid of this entirely to not obfuscate, what's going on inside.
+// @TODO: switch to log-masking everywhere?
 Expr softmax(Expr a, Expr mask) {
-  return Expression<SoftmaxNodeOp>(a, mask);
+  auto logMask = (1 - mask) * -99999999.f;
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_operators.cpp]: It might be a bit more expensive, but I am quite sure it's negligible. Those masks are usually small as they broadcast to batch or time dimension and also computed once per batch. Even it enters this step once for every time step, the hashing algorithm will reuse exactly the same log-mask that was computed once. 

The pain the previous solution was inflicting on me is not worth it :)

File: src/graph/expression_operators.cpp [hunk 9]
```diff
@@ -198,16 +205,32 @@ Expr flatten_2d(Expr a) {
   return Expression<ReshapeNodeOp>(a, shape);
 }
 
-Expr rows(Expr a, const std::vector<size_t>& indices) {
+Expr rows(Expr a, Expr indices) {
   return Expression<RowsNodeOp>(a, indices);
 }
 
-Expr cols(Expr a, const std::vector<size_t>& indices) {
+Expr rows(Expr a, const std::vector<IndexType>& indices) {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_operators.cpp]: Yes. For now. Might be worth getting rid of in the future.

File: src/graph/expression_operators.cpp [hunk 10]
```diff
@@ -198,16 +205,32 @@ Expr flatten_2d(Expr a) {
...
 }
 
-Expr select(Expr a, int axis, const std::vector<size_t>& indices) {
-  return Expression<SelectNodeOp>(a, axis, indices);
+Expr cols(Expr a, const std::vector<IndexType>& indices) {
+  auto indexExpr = a->graph()->indices(indices);
+  return cols(a, indexExpr);
+}
+
+Expr select(Expr a, Expr indices, int axis) {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_operators.cpp]: I believe so.

File: src/graph/expression_operators.h [hunk 11]
```diff
@@ -105,15 +105,21 @@ Expr atleast_nd(Expr a, size_t dims);
...
+Expr select(Expr a, Expr indices, int axis);
+Expr select(Expr a, const std::vector<IndexType>& indices, int axis);
 
 /*********************************************************/
 
 Expr sum(Expr a, int ax = 0);
 
-Expr softmax(Expr a, Expr mask = nullptr);
+Expr softmax(Expr a);
+Expr softmax(Expr a, Expr mask);
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_operators.h]: Done and moved the TODOs to the header.

File: src/graph/expression_operators.h [hunk 12]
```diff
@@ -105,15 +105,21 @@ Expr atleast_nd(Expr a, size_t dims);
...
 
 Expr logsoftmax(Expr a);
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/expression_operators.h]: I am rather thinking of removing the mask parameters from softmax instead. Keeping it for backward compatibility for the moment.

File: src/graph/node_operators.h [hunk 13]
```diff
@@ -9,8 +9,9 @@ namespace marian {
 struct ConstantNode : public Node {
   ConstantNode(Ptr<ExpressionGraph> graph,
                const Shape& shape,
-               const NodeInitializer& init)
-      : Node(graph, shape),  // TODO: add value_type
+               const NodeInitializer& init,
+               Type value_type = Type::float32)
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/node_operators.h]: I am following Numpy. Checked again to be sure, and they match.

File: src/graph/node_operators_binary.h [hunk 14]
```diff
@@ -458,6 +458,113 @@ struct ScalarProductNodeOp : public NaryNodeOp {
...
+  }
+
+  NodeOps backwardOps() override {
+    return {NodeOp(PasteRows(child(0)->grad(), adj_, child(1)->val()))};
+  }
+
+  template <class... Args>
+  Shape newShape(Expr a, size_t num) {
+    Shape shape = a->shape();
+    ABORT_IF(shape.size() != 2,
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/node_operators_binary.h]: That was in there forever. So this does not introduce backwards-incompatibility. If you can point me later to code where this is an issue I can fix this.

File: src/models/transformer.h [hunk 15]
```diff
@@ -178,7 +178,8 @@ class Transformer : public EncoderOrDecoderBase {
 
   void collectOneHead(Expr weights, int dimBeam) {
     // select first head, this is arbitrary as the choice does not really matter
-    auto head0 = select(weights, -3, {0}); // @TODO: implement an index() or slice() operator and use that
+    std::vector<IndexType> idx0 = {0};
+    auto head0 = select(weights, idx0, -3); // @TODO: implement an index() or slice() operator and use that
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/models/transformer.h]: Done.

File: src/tensors/cpu/tensor_operators.cpp [hunk 16]
```diff
@@ -271,27 +271,24 @@ void TransposeNDGrad(Tensor out, Tensor in, const std::vector<int>& vAxis) {
     TransposeGeneric<true>(out, in, vAxis);
 }
 
-void Softmax(Tensor out_, Tensor in_, Tensor mask_) {
+void Softmax(Tensor out_, Tensor in_) {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/tensors/cpu/tensor_operators.cpp]: Don't like them either. I think that was the Intel guy who wrote those CPU kernels. Removed them in a few and added a @TODO to get rid of the rest. I will have to revisit those while I make those function fit for working with different types.

File: src/tensors/cpu/tensor_operators.cpp [hunk 17]
```diff
@@ -382,18 +379,20 @@ void LogSoftmaxGrad(Tensor grad_, Tensor adj_, Tensor val_) {
...
+  size_t rows = indices->size();
 
   float* out = out_->data();
   const float* in = in_->data();
 
 #pragma omp parallel for
   for(size_t j = 0; j < rows; ++j) {
     size_t dst = j;
-    size_t src = indices[j];
+    size_t src = (size_t)indices->data<IndexType>()[j];
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/tensors/cpu/tensor_operators.cpp]: Added a @todo and will think about this when I write the type-aware kernels.

File: src/tensors/cpu/tensor_operators.cpp [hunk 18]
```diff
@@ -404,15 +403,18 @@ void CopyRows(Tensor out_,
 
 void PasteRows(Tensor out_,
                const Tensor in_,
-               const std::vector<size_t>& indices) {
+               const Tensor indices) {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/tensors/cpu/tensor_operators.cpp]: As above.

File: src/tensors/cpu/tensor_operators.cpp [hunk 19]
```diff
@@ -440,17 +445,20 @@ void CopyCols(Tensor out_,
     float* rowOut = out + j * colsOut;
 
     for(size_t i = 0; i < colsOut; ++i) {
-      rowOut[i] = rowIn[indices[i]];
+      rowOut[i] = rowIn[indices->data<IndexType>()[i]];
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/tensors/cpu/tensor_operators.cpp]: As above.

File: src/tensors/cpu/tensor_operators.cpp (hunk 1 above)
Comment (emjotde) [reply src/tensors/cpu/tensor_operators.cpp]: Brrr. As above.

File: src/tensors/gpu/tensor_operators.cu (hunk 2 above)
Comment (emjotde) [reply src/tensors/gpu/tensor_operators.cu]: Copied over comment from Node.

File: src/graph/node_operators_binary.h [hunk 20]
```diff
@@ -458,6 +458,113 @@ struct ScalarProductNodeOp : public NaryNodeOp {
...
+
+  const std::string type() override { return "rows"; }
+
+  const std::string color() override { return "orange"; }
+};
+
+struct SelectNodeOp : public NaryNodeOp {
+  SelectNodeOp(Expr a, Expr indices, int axis)
+      : NaryNodeOp({a, indices}, newShape(a, axis, indices->shape().elements())),
+        axis_{a->shape().axis(axis)} {
```
[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/node_operators_binary.h]: Can you give me a minimal example that you expect to not work?

[the remark this answers is not in the document]
Comment (emjotde) [reply src/graph/node_operators_binary.h]: OK, you comment refers to an old version of select that was broken. Current version works and we could do the following:

```c++
Expr rows(Expr a, Expr indices) {
  // @TODO:: replace with `select(a, indices, -2)` 
  // as soon as select is efficient enough
  return Expression<RowsNodeOp>(a, indices);
}

Expr cols(Expr a, Expr indices) {
  // @TODO:: replace with `select(a, indices, -1)` 
  // as soon as select is efficient enough
  return Expression<ColsNodeOp>(a, indices);
}
```

Checked, passes all regression tests. Just not very efficient at the moment as it performs unnecessary computations per element.

**Comment by kpu** [comment]
@frankseide
>     * Why int32_t and not size_t? That should be the natural type.
According to https://devtalk.nvidia.com/default/topic/389324/64-bit-integer-operations/ NVIDIA GPUs do not have native 64-bit integer multiply but emulate it in multiple instructions.  So there is a speed reason to use the GPU's native 32-bit type.
